Index: class.krumo.php
===================================================================
--- class.krumo.php	(revision 49021)
+++ class.krumo.php	(working copy)
@@ -60,6 +60,8 @@
 */
 Class krumo {
 
+	static $objectMarkers = array();
+
 	/**
 	* Return Krumo version
 	*
@@ -610,14 +612,9 @@
 <?php
 		// flee the hive
 		//
-		$_recursion_marker = krumo::_marker();
 		if ($hive =& krumo::_hive($dummy)) {
 			foreach($hive as $i=>$bee){
-				if (is_object($bee)) {
-					unset($hive[$i]->$_recursion_marker);
-					} else {
-					unset($hive[$i][$_recursion_marker]);
-					}
+				krumo::unsetMarker($hive[$i]);
 				}
 			}
 
@@ -925,10 +922,7 @@
 			
 			// stain it
 			//
-			$_recursion_marker = krumo::_marker();
-			(is_object($bee))
-				? @($bee->$_recursion_marker++)
-				: @($bee[$_recursion_marker]++);
+			krumo::setMarker($bee);
 			
 			$_[0][] =& $bee;
 			}
@@ -955,9 +949,7 @@
 		// prevent endless recursion loops
 		//
 		$_recursion_marker = krumo::_marker();
-		$_r = ($_is_object)
-			? @$data->$_recursion_marker
-			: @$data[$_recursion_marker] ;
+		$_r = krumo::getMarker($data);
 		$_r = (integer) $_r;
 
 		// recursion detected
@@ -1288,7 +1280,36 @@
 		}
 
 	// -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
+	
+	private static function setMarker(&$obj) {
+		if (is_object($obj)) {
+			krumo::$objectMarkers[spl_object_hash($obj)] = 1;
+			} else {
+			$_recursion_marker = krumo::_marker();
+			$obj[$_recursion_marker] = 1;
+			}
+		}
 
+	private static function getMarker(&$obj) {
+		if (is_object($obj)) {
+			return isset(krumo::$objectMarkers[spl_object_hash($obj)])
+				? krumo::$objectMarkers[spl_object_hash($obj)] : null;
+			} else {
+			$_recursion_marker = krumo::_marker();
+			return isset($obj[$_recursion_marker])
+				? $obj[$_recursion_marker] : null;
+			}
+		}
+
+	private static function unsetMarker(&$obj) {
+		if (is_object($obj)) {
+			unset(krumo::$objectMarkers[spl_object_hash($obj)]);
+			} else {
+			$_recursion_marker = krumo::_marker();
+			unset($obj[$_recursion_marker]);
+			}
+		}
+
 //--end-of-class--	
 }
 
